src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -750,5 +733,311 @@ private static string ToHtmlCase(string name)
         {
             return HtmlCaseRegex.Replace(name, HtmlCaseRegexReplacement).ToLowerInvariant();
         }
+
+        // Internal for testing
+        internal class RequiredAttributeParser
+        {
+            private static readonly char[] InvalidPlainAttributeNameCharacters = { ' ', '\t', ',', RequiredAttributeWildcardSuffix };
+            private static readonly char[] InvalidCSSAttributeNameCharacters = (new[] { ' ', '\t', ',', ']' }).Concat(TagHelperRequiredAttributeDescriptor.SupportedCSSValueOperators).ToArray();
+            private static readonly char[] InvalidCSSQuotelessValueCharacters = { ' ', '\t', ']' };
+
+            private int _index;
+            private string _requiredAttributes;
+
+            public RequiredAttributeParser(string requiredAttributes)
+            {
+                _requiredAttributes = requiredAttributes;
+            }
+
+            private char Current => _requiredAttributes[_index];
+
+            private bool AtEnd => _index >= _requiredAttributes.Length;
+
+            public bool TryParse(
+                ErrorSink errorSink,
+                out IEnumerable<TagHelperRequiredAttributeDescriptor> requiredAttributeDescriptors)
+            {
+                if (string.IsNullOrEmpty(_requiredAttributes))
+                {
+                    requiredAttributeDescriptors = Enumerable.Empty<TagHelperRequiredAttributeDescriptor>();
+                    return true;
+                }
+
+                requiredAttributeDescriptors = null;
+                var descriptors = new List<TagHelperRequiredAttributeDescriptor>();
+
+                while (!AtEnd)
+                {
+                    PassOptionalWhitespace();
+
+                    TagHelperRequiredAttributeDescriptor descriptor;
+                    if (At('['))
+                    {
+                        descriptor = ParseCSSSelector(errorSink);
+                    }
+                    else
+                    {
+                        descriptor = ParsePlainSelector(errorSink);
```

Comment:
If the string ends with a comma or a comma then whitespace, `ParsePlainSelector()` will be called and complain about the name rather than the trailing comma. My suggestion is to ensure `!AtEnd` before trying to parse the plain name.